test/test.js

```diff
@@ -33,11 +33,29 @@ describe('Routes', () => {
       assert.strictEqual(await res.text(), 'Hello World!')
     })
   })
-  describe('GET /retrieval', () => {
-    it('responds', async () => {
-      const res = await fetch(`${spark}/retrieval`)
+  describe('PUT /retrieval', () => {
+    it('creates a retrieval', async () => {
+      const res = await fetch(`${spark}/retrieval`, { method: 'PUT' })
       assert.strictEqual(res.status, 200)
-      assert.deepStrictEqual(await res.json(), { hello: 'world' })
+      const body = await res.json()
+      assert.strictEqual(typeof body.id, 'number')
+      assert.strictEqual(typeof body.cid, 'string')
+      assert.strictEqual(typeof body.providerAddress, 'string')
+      assert.strictEqual(typeof body.protocol, 'string')
+    })
+    it('uses random retrieval templates', async () => {
+      let lastCid
+      for (let i = 0; i < 100; i++) {
+        const res = await fetch(`${spark}/retrieval`, { method: 'PUT' })
+        assert.strictEqual(res.status, 200)
+        const body = await res.json()
+        if (!lastCid) {
+          lastCid = body.cid
+        } else if (lastCid !== body.cid) {
+          return
+        }
+      }
+      throw new Error('All CIDs were the same')
```

Comment:
1. I find it quite challenging to understand how exactly the assertion works in this test.

2. The error message seems misleading - we don't know if _ALL_ CIDs are the same, only that _SOME TWO CONSECUTIVE_ CIDs are the same.

3. You are running this loop 100 times, but the database has only 2 rows to choose from. I don't see how we can generate 100 unique CIDs from those 2 rows? Maybe the test passes because the server returns CID1 - CID2 - CID1 - CID2 and we never get two consecutive responses with the same CID. But that's not really correct, right?

I am proposing to make the following changes:
1. Run the loop fewer times. If we have only 2 records then we must run it twice. With more records in the DB, I would choose 10 iterations.
2. Collect all CIDs into an array.
3. At the end, assert that this array contains unique items only. If the assertion fails, then print the entire array to make it easier to understand what went wrong.
  
   To detect repeated items, we can e.g. sort the array and then check for duplicates using your algorithm.

Reply:
It tries 100 attempts to find two different CIDs. I thought the iteration count of 100 is safe enough to not have to think about it. We can also lower it to 50 or less if you prefer?

>  I don't see how we can generate 100 unique CIDs from those 2 rows?

We are never generating CIDs, I don't think we could? There is a fixed list of CIDs, stored in `retrieval_templates`. On every request, we are generating a `retrieval`, based on one of the `retrieval_templates`. This test ensures that `retrieval_templates` are chosen at random. Should I add a comment to make this more clear?

Reply:
> It tries 100 attempts to find two different CIDs. I thought the iteration count of 100 is safe enough to not have to think about it. We can also lower it to 50 or less if you prefer?
> 
> > I don't see how we can generate 100 unique CIDs from those 2 rows?
> 
> We are never generating CIDs, I don't think we could? There is a fixed list of CIDs, stored in `retrieval_templates`. On every request, we are generating a `retrieval`, based on one of the `retrieval_templates`. This test ensures that `retrieval_templates` are chosen at random. Should I add a comment to make this more clear?

Sorry for not being clear. By generating CIDs I meant generating retrieval jobs by picking random CIDs.

I see that you have another PR to add more than 2 CIDs to pick from, that resolves one of my concerns.

I am still confused about how is your test supposed to work.
- You store the first CID fetched inside `lastCid` (how confusing!) and then run another requests, up to 99.
- After the second request, if `lastCid !== body.cid` then you `return`, therefore terminating the entire test with success.
- If the second request returns the same CID as the one before, then you keep trying.

I think I am starting to understand it now, but it took me a while. How about rewriting the test along the following lines?

```js
const pickRandomCid = async () {
  const res = await fetch(`${spark}/retrieval`, { method: 'POST' })
  assert.strictEqual(res.status, 200)
  const body = await res.json()
  return body.cid
}

const firstCid = await pickRandomCid()

for (let i = 0; i < 100; i++) {
  let anotherCid = await pickRandomCid()
  if (anotherCid !== firstCid) {
    // The service seems to return different CIDs 
    // for different requests - the test passed.
    return
  }
}

throw new Error(`All requests returned the same CID: ${firstCid}.`)
```

Reply:
This was really helpful to read! Thank you for sharing. I refactored the test very similar to what you proposed 🙏